### ISSUE TITLE
[P1] feat(attention): Add softcapping support for Gemma 2 and other models

### ISSUE THREAD

**Issue by noahgift** [paiml/realizar#27]
## Summary

Add attention logit softcapping to the attention implementations. Softcapping applies `softcap * tanh(logits / softcap)` to attention scores before softmax, preventing extreme values and improving training stability.

## Motivation

Gemma 2 and other recent models use attention softcapping (typically softcap=50.0). Without this, realizar cannot correctly run these models - the attention distribution becomes incorrect.

Candle supports softcapping in both Flash Attention and Metal SDPA kernels.

### Simple Addition to Existing Attention

```rust
/// Apply softcapping: softcap * tanh(x / softcap)
/// Bounds output to [-softcap, +softcap]
#[inline]
pub fn softcap(x: f32, cap: f32) -> f32 {
    cap * (x / cap).tanh()
}

/// Vectorized softcap for SIMD
pub fn softcap_vec(x: &mut [f32], cap: f32) {
    let inv_cap = 1.0 / cap;
    for v in x.iter_mut() {
        *v = cap * (*v * inv_cap).tanh();
    }
}
```

### Integration with Attention

```rust
pub struct AttentionConfig {
    pub num_heads: usize,
    pub head_dim: usize,
    pub scale: f32,
    pub softcap: Option<f32>,  // NEW: None = no capping
    // ...
}

impl Attention {
    pub fn forward(&self, q: &Tensor, k: &Tensor, v: &Tensor) -> Result<Tensor> {
        // Q @ K^T
        let mut scores = matmul(q, &k.transpose(-2, -1)?)?;
        
        // Scale
        scores = scores.mul_scalar(self.config.scale)?;
        
        // NEW: Apply softcap if configured
        if let Some(cap) = self.config.softcap {
            scores = scores.apply(|x| softcap(x, cap))?;
        }
        
        // Causal mask (if needed)
        if self.config.causal {
            scores = apply_causal_mask(scores)?;
        }
        
        // Softmax
        let attn_weights = softmax(&scores, -1)?;
        
        // @ V
        matmul(&attn_weights, v)
    }
}
```

### Flash Attention Integration

```rust
impl FlashAttention {
    pub fn forward_with_softcap(
        &self,
        q: &Tensor,
        k: &Tensor, 
        v: &Tensor,
        softcap: Option<f32>,
    ) -> Result<Tensor> {
        // In block-wise computation, apply softcap after Q@K^T scaling
        // but before online softmax accumulation
        
        for block in blocks {
            let mut block_scores = q_block @ k_block.T * scale;
            
            // Apply softcap within the block
            if let Some(cap) = softcap {
                block_scores = softcap_tensor(&block_scores, cap);
            }
            
            // Continue with online softmax...
        }
    }
}
```

### GPU Kernel (trueno-gpu)

```rust
// Fused scale + softcap kernel
pub fn generate_softcap_kernel() -> String {
    r#"
    .visible .entry softcap_kernel(
        .param .u64 input,
        .param .u64 output,
        .param .f32 scale,
        .param .f32 softcap,
        .param .u64 n
    ) {
        // x = scale * input[i]
        // output[i] = softcap * tanh(x / softcap)
        
        // tanh approximation for speed:
        // tanh(x) ≈ x * (27 + x²) / (27 + 9x²) for |x| < 3
        // or use __expf for full precision
    }
    "#
}
```

## Models Requiring Softcapping

| Model | Softcap Value |
|-------|---------------|
| Gemma 2 (2B, 9B, 27B) | 50.0 |
| Gemma 2 variants | 50.0 |

## Acceptance Criteria

- [ ] `softcap()` function for scalar values
- [ ] `softcap_tensor()` for tensor application
- [ ] `AttentionConfig.softcap: Option<f32>` field
- [ ] Standard attention with softcap support
- [ ] Flash Attention v1/v2 with softcap support
- [ ] SIMD-accelerated softcap (AVX2, NEON)
- [ ] GPU kernel via trueno-gpu
- [ ] GGUF metadata parsing for softcap value
- [ ] Unit tests with known Gemma 2 outputs
- [ ] Benchmark overhead (should be <5% latency increase)

## References

- Gemma 2 Technical Report: https://arxiv.org/abs/2408.00118
- Candle `scaled_dot_product_attention.metal` (softcapping implementation)
- Candle Flash Attention softcap parameter

**Comment by noahgift** [comment]
## Triage Update (2026-02-21)

**Status: NOT STARTED**

Zero results for `softcap` in source code (excluding docs/tests). No softcap parameter in any attention kernel.
